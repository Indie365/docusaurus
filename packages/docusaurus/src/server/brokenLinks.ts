/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import _ from 'lodash';
import logger from '@docusaurus/logger';
import {matchRoutes} from 'react-router-config';
import {parseURLPath, serializeURLPath, type URLPath} from '@docusaurus/utils';
import {getAllFinalRoutes} from './utils';
import type {RouteConfig, ReportingSeverity} from '@docusaurus/types';

type BrokenLink = {
  link: string;
  resolvedLink: string;
  anchor: boolean;
};

type BrokenLinksByLocation = {[location: string]: BrokenLink[]};

type CollectedLinks = {
  [location: string]: {links: string[]; anchors: string[]};
};

function getPageBrokenLinks({
  allCollectedLinks,
  pagePath,
  pageLinks,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  pagePath: string;
  pageLinks: string[];
  pageAnchors: string[];
  routes: RouteConfig[];
}): BrokenLink[] {
  // console.log('routes:', routes);
  function isPathBrokenLink(linkPath: URLPath) {
    const matchedRoutes = [linkPath.pathname, decodeURI(linkPath.pathname)]
      // @ts-expect-error: React router types RouteConfig with an actual React
      // component, but we load route components with string paths.
      // We don't actually access component here, so it's fine.
      .map((l) => matchRoutes(routes, l))
      .flat();
    return matchedRoutes.length === 0;
  }

  function isAnchorBrokenLink(linkPath: URLPath) {
    const {pathname, hash} = linkPath;

    // Link has no hash: it can't be a broken anchor link
    if (hash === undefined) {
      return false;
    }

    const targetPage =
      allCollectedLinks[pathname] || allCollectedLinks[decodeURI(pathname)];

    // link with anchor to a page that does not exist (or did not collect any
    // link/anchor) is considered as a broken anchor
    if (!targetPage) {
      return true;
    }

    // it's a broken anchor if the target page exists
    // but the anchor does not exist on that page
    return !targetPage.anchors.includes(hash);
  }

  const brokenLinks = pageLinks.flatMap((link) => {
    const linkPath = parseURLPath(link, pagePath);
    if (isPathBrokenLink(linkPath)) {
      return [
        {
          link,
          resolvedLink: serializeURLPath(linkPath),
          anchor: false,
        },
      ];
    }
    if (isAnchorBrokenLink(linkPath)) {
      return [
        {
          link,
          resolvedLink: serializeURLPath(linkPath),
          anchor: true,
        },
      ];
    }
    return [];
  });

  return brokenLinks;
}

/**
 * The route defs can be recursive, and have a parent match-all route. We don't
 * want to match broken links like /docs/brokenLink against /docs/*. For this
 * reason, we only consider the "final routes" that do not have subroutes.
 * We also need to remove the match-all 404 route
 */
function filterIntermediateRoutes(routesInput: RouteConfig[]): RouteConfig[] {
  const routesWithout404 = routesInput.filter((route) => route.path !== '*');
  return getAllFinalRoutes(routesWithout404);
}

function getAllBrokenLinks({
  allCollectedLinks,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  routes: RouteConfig[];
}): {brokenLinks: BrokenLinksByLocation; brokenAnchors: BrokenLinksByLocation} {
  const filteredRoutes = filterIntermediateRoutes(routes);

  const allBrokenLinks = _.mapValues(
    allCollectedLinks,
    (pageCollectedData, pagePath) =>
      getPageBrokenLinks({
        allCollectedLinks,
        pageLinks: pageCollectedData.links,
        pageAnchors: pageCollectedData.anchors,
        pagePath,
        routes: filteredRoutes,
      }),
  );

  function splitBrokenLinks(collect: {[x: string]: BrokenLink[]}) {
    const brokenLinks: {[x: string]: BrokenLink[]} = {};
    const brokenAnchors: {[x: string]: BrokenLink[]} = {};

    Object.entries(collect).forEach(([page, links]) => {
      const [linksFiltered, anchorsFiltered] = _.partition(
        links,
        (link) => link.anchor === false,
      );

      if (linksFiltered.length > 0) {
        brokenLinks[page] = linksFiltered;
      }
      if (anchorsFiltered.length > 0) {
        brokenAnchors[page] = anchorsFiltered;
      }
    });

    return {brokenLinks, brokenAnchors};
  }

  return splitBrokenLinks(allBrokenLinks);
}

function brokenLinkMessage(brokenLink: BrokenLink): string {
  const showResolvedLink = brokenLink.link !== brokenLink.resolvedLink;
  return `${brokenLink.link}${
    showResolvedLink ? ` (resolved as: ${brokenLink.resolvedLink})` : ''
  }`;
}

function createBrokenLinksMessage(
  pagePath: string,
  allBrokenLinks: BrokenLink[],
): string {
  const type = allBrokenLinks[0]?.anchor === true ? 'anchor' : 'link';

  const anchorMessage =
    allBrokenLinks.length > 0
      ? `- Broken ${type} on source page path = ${pagePath}:
   -> linking to ${allBrokenLinks
     .map(brokenLinkMessage)
     .join('\n   -> linking to ')}`
      : '';

  return `${anchorMessage}`;
}

function getAnchorBrokenLinksErrorMessage(allBrokenLinks: {
  [location: string]: BrokenLink[];
}): string | undefined {
  if (Object.keys(allBrokenLinks).length === 0) {
    return undefined;
  }

  return `Docusaurus found broken anchors!

Please check the pages of your site in the list below, and make sure you don't reference any anchor that does not exist.
Note: it's possible to ignore broken anchors with the 'onBrokenAnchors' Docusaurus configuration, and let the build pass.

Exhaustive list of all broken anchors found:
${Object.entries(allBrokenLinks)
  .map(([pagePath, brokenLinks]) =>
    createBrokenLinksMessage(pagePath, brokenLinks),
  )
  .join('\n')}
`;
}

function getPathBrokenLinksErrorMessage(allBrokenLinks: {
  [location: string]: BrokenLink[];
}): string | undefined {
  if (Object.keys(allBrokenLinks).length === 0) {
    return undefined;
  }

  /**
   * If there's a broken link appearing very often, it is probably a broken link
   * on the layout. Add an additional message in such case to help user figure
   * this out. See https://github.com/facebook/docusaurus/issues/3567#issuecomment-706973805
   */
  function getLayoutBrokenLinksHelpMessage() {
    const flatList = Object.entries(allBrokenLinks).flatMap(
      ([pagePage, brokenLinks]) =>
        brokenLinks.map((brokenLink) => ({pagePage, brokenLink})),
    );

    const countedBrokenLinks = _.countBy(
      flatList,
      (item) => item.brokenLink.link,
    );

    const FrequencyThreshold = 5; // Is this a good value?
    const frequentLinks = Object.entries(countedBrokenLinks)
      .filter(([, count]) => count >= FrequencyThreshold)
      .map(([link]) => link);

    if (frequentLinks.length === 0) {
      return '';
    }

    return logger.interpolate`

It looks like some of the broken links we found appear in many pages of your site.
Maybe those broken links appear on all pages through your site layout?
We recommend that you check your theme configuration for such links (particularly, theme navbar and footer).
Frequent broken links are linking to:${frequentLinks}`;
  }

  return `Docusaurus found broken links!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.${getLayoutBrokenLinksHelpMessage()}

Exhaustive list of all broken links found:
${Object.entries(allBrokenLinks)
  .map(([pagePath, brokenLinks]) =>
    createBrokenLinksMessage(pagePath, brokenLinks),
  )
  .join('\n')}
`;
}

export async function handleBrokenLinks({
  allCollectedLinks,
  onBrokenLinks,
  onBrokenAnchors,
  routes,
}: {
  allCollectedLinks: CollectedLinks;
  onBrokenLinks: ReportingSeverity;
  onBrokenAnchors: ReportingSeverity;
  routes: RouteConfig[];
}): Promise<void> {
  if (onBrokenLinks === 'ignore' && onBrokenAnchors === 'ignore') {
    return;
  }

  const {brokenLinks, brokenAnchors} = getAllBrokenLinks({
    allCollectedLinks,
    routes,
  });

  const pathErrorMessage = getPathBrokenLinksErrorMessage(brokenLinks);
  if (pathErrorMessage) {
    logger.report(onBrokenLinks)(pathErrorMessage);
  }

  const anchorErrorMessage = getAnchorBrokenLinksErrorMessage(brokenAnchors);
  if (anchorErrorMessage) {
    logger.report(onBrokenAnchors)(anchorErrorMessage);
  }
}
