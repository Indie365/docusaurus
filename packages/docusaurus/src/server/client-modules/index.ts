/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Plugin} from '@docusaurus/types';

export function loadClientModules(plugins: Plugin<any>[]): string[] {
  return plugins
    .map(plugin =>
      !plugin.getClientModules ? null : plugin.getClientModules(),
    )
    .reduce((a, b) => a.concat(b), [])
    .filter(Boolean);
}
