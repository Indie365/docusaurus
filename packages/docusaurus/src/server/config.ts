/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import fs from 'fs-extra';
import logger from '@docusaurus/logger';
import {
  DEFAULT_CONFIG_FILE_NAME,
  DEFAULT_THEME_FILE_NAME,
  findAsyncSequential,
  loadFreshModule,
} from '@docusaurus/utils';
import {validateConfig} from './configValidation';
import type {DocusaurusConfig, LoadContext} from '@docusaurus/types';

async function getConventionalSiteConfigPath(siteDir: string) {
  // We could support .mjs, .ts, etc. in the future
  const candidates = ['.ts', '.mts', '.cts', '.js', '.mjs', '.cjs'].map(
    (ext) => DEFAULT_CONFIG_FILE_NAME + ext,
  );
  const configPath = await findAsyncSequential(
    candidates.map((file) => path.join(siteDir, file)),
    fs.pathExists,
  );
  if (!configPath) {
    const errorMessage = logger.interpolate`No config file found.
Expected one of:${candidates}
You can provide a custom config path with the code=${'--config'} option.`;
    throw new Error(errorMessage);
  }
  return configPath;
}

export async function loadSiteConfig({
  siteDir,
  customConfigFilePath,
}: {
  siteDir: string;
  customConfigFilePath?: string;
}): Promise<Pick<LoadContext, 'siteConfig' | 'siteConfigPath'>> {
  const siteConfigPath = customConfigFilePath
    ? path.resolve(siteDir, customConfigFilePath)
    : await getConventionalSiteConfigPath(siteDir);

  if (!(await fs.pathExists(siteConfigPath))) {
    throw new Error(`Config file at "${siteConfigPath}" not found.`);
  }

  const importedConfig = await loadFreshModule(siteConfigPath);

  const loadedConfig: unknown =
    typeof importedConfig === 'function'
      ? await importedConfig()
      : await importedConfig;

  const siteConfig = validateConfig(
    loadedConfig,
    path.relative(siteDir, siteConfigPath),
  );
  return {siteConfig, siteConfigPath};
}

async function findConventionalThemeConfigPath(
  siteDir: string,
): Promise<string | undefined> {
  // We could support .mjs, .ts, etc. in the future
  const candidates = ['.tsx', '.ts', '.jsx', '.js'].map(
    (ext) => DEFAULT_THEME_FILE_NAME + ext,
  );
  const themeConfigPath = await findAsyncSequential(
    candidates.map((file) => path.join(siteDir, file)),
    fs.pathExists,
  );

  return themeConfigPath;
}

// TODO add tests for this
export async function findThemeConfigPath(
  siteDir: string,
  siteConfig: DocusaurusConfig,
): Promise<string | undefined> {
  // TODO add support for custom themeConfig file path
  // EX: siteConfig.themeConfig: './theme.tsx'
  // Note: maybe it would be simpler to provide this path through the CLI?
  if (typeof siteConfig.themeConfig === 'string') {
    const customThemeConfigPath = siteConfig.themeConfig;
    if (!(await fs.pathExists(customThemeConfigPath))) {
      throw new Error(
        `Theme config file at "${customThemeConfigPath}" not found.`,
      );
    }
    return customThemeConfigPath;
  }
  const conventionalThemeConfigPath = await findConventionalThemeConfigPath(
    siteDir,
  );
  // In Docusaurus v3 we require users to provide either the theme config
  // through the conventional path, or through the legacy siteConfig attribute
  // To avoid issues we prevent users to not provide any theme config at all
  if (
    !conventionalThemeConfigPath &&
    Object.keys(siteConfig.themeConfig ?? {}).length === 0
  ) {
    throw new Error(
      `Theme config file couldn't be found at ${DEFAULT_THEME_FILE_NAME}.js or ${DEFAULT_THEME_FILE_NAME}.tsx`,
    );
  }
  return conventionalThemeConfigPath;
}
