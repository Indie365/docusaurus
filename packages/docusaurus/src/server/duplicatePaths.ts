/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import chalk from 'chalk';
import {RouteConfig, OnDuplicateRoutes} from '@docusaurus/types';

export function getAllDuplicateRoutes(
  pluginsRouteConfigs: RouteConfig[],
): string[] {
  const routesSeen: Record<string, any> = {};
  const duplicateRoutes: string[] = [];
  function getDuplicateRoutes(routeConfigs: RouteConfig[]): string[] {
    for (let i = 0; i < routeConfigs.length; i += 1) {
      const routeConfig = routeConfigs[i];
      if (routesSeen.hasOwnProperty(routeConfig.path)) {
        duplicateRoutes.push(routeConfig.path);
      } else {
        routesSeen[routeConfig.path] = true; // dummy value for object
      }
      if (routeConfig.routes !== undefined) {
        getDuplicateRoutes(routeConfig.routes);
      }
    }
    return duplicateRoutes;
  }
  return getDuplicateRoutes(pluginsRouteConfigs);
}

export function getDuplicateRoutesMessage(
  allDuplicateRoutes: string[],
): string {
  const message = allDuplicateRoutes
    .map(
      (duplicateRoute) =>
        `Attempting to create page at ${duplicateRoute}, but a page already exists at this route`,
    )
    .join('\n');
  return message;
}

export function handleDuplicateRoutes(
  pluginsRouteConfigs: RouteConfig[],
  onDuplicateRoutes: OnDuplicateRoutes,
): void {
  if (onDuplicateRoutes === 'ignore') {
    return;
  }
  const duplicatePaths: string[] = getAllDuplicateRoutes(pluginsRouteConfigs);
  const message: string = getDuplicateRoutesMessage(duplicatePaths);
  if (message) {
    const finalMessage = `Duplicate routes found!\n${message}\nThis could lead to non-deterministic routing behavior`;
    if (onDuplicateRoutes === 'log') {
      console.log(chalk.blue(finalMessage));
    } else if (onDuplicateRoutes === 'warn') {
      console.warn(chalk.yellow(finalMessage));
    } else if (onDuplicateRoutes === 'throw') {
      throw new Error(finalMessage);
    }
  }
}

export function warnAboutOverridingRoutes(
  pluginsRouteConfigs: RouteConfig[],
): void {
  // Accumulate all the routes by recursively exploring each RouteConfig
  const routesAccumulator: string[] = [];
  function getAllRoutes(routeConfigs: RouteConfig[]): string[] {
    for (let i = 0; i < routeConfigs.length; i += 1) {
      const routeConfig = routeConfigs[i];
      routesAccumulator.push(routeConfig.path);
      if (routeConfig.routes !== undefined) {
        getAllRoutes(routeConfig.routes);
      }
    }
    return routesAccumulator;
  }

  const allRoutes = getAllRoutes(pluginsRouteConfigs);

  // Sort the allRoutes array in lexicographical order
  // Then check if each route is equal to the next route
  // If yes, one of these routes will be overridden so we warn the user
  allRoutes.sort((a, b) => a.localeCompare(b));
  for (let i = 0; i < allRoutes.length - 1; i += 1) {
    if (allRoutes[i] === allRoutes[i + 1]) {
      console.warn(
        `${
          chalk.yellow(`warning `) + chalk.bold.yellow(`Routes Override: `)
        }Attempting to create page at "${
          allRoutes[i]
        }" but a page already exists at this path\n${chalk.bold.yellow(
          `This could lead to non-deterministic routing behavior`,
        )}`,
      );
    }
  }
}
