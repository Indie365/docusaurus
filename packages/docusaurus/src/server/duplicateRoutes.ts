/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import chalk from 'chalk';
import {RouteConfig, OnDuplicateRoutes} from '@docusaurus/types';

export function getAllDuplicateRoutes(
  pluginsRouteConfigs: RouteConfig[],
): string[] {
  const routesSeen: Record<string, any> = {};
  const duplicateRoutes: string[] = [];
  function getDuplicateRoutes(routeConfigs: RouteConfig[]): string[] {
    for (let i = 0; i < routeConfigs.length; i += 1) {
      const routeConfig = routeConfigs[i];
      if (routesSeen.hasOwnProperty(routeConfig.path)) {
        duplicateRoutes.push(routeConfig.path);
      } else {
        routesSeen[routeConfig.path] = true; // dummy value
      }
      if (routeConfig.routes !== undefined) {
        getDuplicateRoutes(routeConfig.routes);
      }
    }
    return duplicateRoutes;
  }
  return getDuplicateRoutes(pluginsRouteConfigs);
}

export function getDuplicateRoutesMessage(
  allDuplicateRoutes: string[],
): string {
  const message = allDuplicateRoutes
    .map(
      (duplicateRoute) =>
        `Attempting to create page at ${duplicateRoute}, but a page already exists at this route`,
    )
    .join('\n');
  return message;
}

export function handleDuplicateRoutes(
  pluginsRouteConfigs: RouteConfig[],
  onDuplicateRoutes: OnDuplicateRoutes,
): void {
  if (onDuplicateRoutes === 'ignore') {
    return;
  }
  const duplicatePaths: string[] = getAllDuplicateRoutes(pluginsRouteConfigs);
  const message: string = getDuplicateRoutesMessage(duplicatePaths);
  if (message) {
    const finalMessage = `Duplicate routes found!\n${message}\nThis could lead to non-deterministic routing behavior`;
    if (onDuplicateRoutes === 'log') {
      console.log(chalk.blue(finalMessage));
    } else if (onDuplicateRoutes === 'warn') {
      console.warn(chalk.yellow(finalMessage));
    } else if (onDuplicateRoutes === 'throw') {
      throw new Error(finalMessage);
    } else {
      // should not reach here if validation is done correctly
      throw new Error(
        'unexpected onDuplicateRoutes value: ${onDuplicateRoutes}',
      );
    }
  }
}
