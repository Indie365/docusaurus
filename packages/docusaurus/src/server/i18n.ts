/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {I18n, DocusaurusConfig, I18nLocaleConfig} from '@docusaurus/types';
import path from 'path';
import {normalizeUrl} from '@docusaurus/utils';
import {getLangDir} from 'rtl-detect';
import logger from '@docusaurus/logger';

function getDefaultLocaleLabel(locale: string) {
  const languageName = new Intl.DisplayNames(locale, {type: 'language'}).of(
    locale,
  );
  return (
    languageName.charAt(0).toLocaleUpperCase(locale) + languageName.substring(1)
  );
}

export function getDefaultLocaleConfig(locale: string): I18nLocaleConfig {
  return {
    label: getDefaultLocaleLabel(locale),
    direction: getLangDir(locale),
    htmlLang: locale,
    baseUrl: undefined,
  };
}

export function shouldWarnAboutNodeVersion(
  version: number,
  locales: string[],
): boolean {
  const isOnlyEnglish = locales.length === 1 && locales.includes('en');
  const isOlderNodeVersion = version < 14;
  return isOlderNodeVersion && !isOnlyEnglish;
}

export async function loadI18n(
  config: DocusaurusConfig,
  options: {locale?: string} = {},
): Promise<I18n> {
  const {i18n: i18nConfig} = config;

  const currentLocale = options.locale ?? i18nConfig.defaultLocale;

  if (!i18nConfig.locales.includes(currentLocale)) {
    logger.warn`The locale name=${currentLocale} was not found in your site configuration: Available locales are: ${i18nConfig.locales}
Note: Docusaurus only support running one locale at a time.`;
  }

  const locales = i18nConfig.locales.includes(currentLocale)
    ? i18nConfig.locales
    : (i18nConfig.locales.concat(currentLocale) as [string, ...string[]]);

  function getLocaleConfig(locale: string): I18nLocaleConfig {
    return {
      ...getDefaultLocaleConfig(locale),
      ...i18nConfig.localeConfigs[locale],
    };
  }

  const localeConfigs = locales.reduce(
    (acc, locale) => ({...acc, [locale]: getLocaleConfig(locale)}),
    {},
  );

  return {
    defaultLocale: i18nConfig.defaultLocale,
    locales,
    currentLocale,
    localeConfigs,
  };
}

function shouldLocalizePath({
  i18n,
  options = {},
}: {
  i18n: I18n;
  options?: {localizePath?: boolean};
}): boolean {
  if (options.localizePath === true || options.localizePath === false) {
    return options.localizePath;
  }
  if (
    i18n.currentLocale !== i18n.defaultLocale ||
    i18n.localeConfigs[i18n.currentLocale].baseUrl
  ) {
    return true;
  }
  return false;
}

export function localizePath({
  pathType,
  path: originalPath,
  i18n,
  options = {},
}: {
  pathType: 'fs' | 'url';
  path: string;
  i18n: I18n;
  options?: {localizePath?: boolean};
}): string {
  if (shouldLocalizePath({i18n, options})) {
    // FS paths need special care, for Windows support
    const localePath =
      i18n.localeConfigs[i18n.currentLocale].baseUrl ?? i18n.currentLocale;
    if (pathType === 'fs') {
      return path.join(originalPath, path.sep, localePath, path.sep);
    }
    // Url paths
    else if (pathType === 'url') {
      return normalizeUrl([originalPath, '/', localePath, '/']);
    }
    // should never happen
    else {
      throw new Error(`Unhandled path type "${pathType}".`);
    }
  } else {
    return originalPath;
  }
}
