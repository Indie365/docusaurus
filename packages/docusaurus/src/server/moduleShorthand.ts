/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export function resolveModuleName(
  moduleName: string,
  moduleRequire: NodeRequire,
  moduleType: 'preset' | 'theme' | 'plugin',
): string {
  const modulePatterns = [
    moduleName,
    `@docusaurus/${moduleType}-${moduleName}`,
    `docusaurus-${moduleType}-${moduleName}`,
    `@${moduleName}/docusaurus-${moduleType}`,
  ];
  // eslint-disable-next-line no-restricted-syntax
  for (const module of modulePatterns) {
    try {
      moduleRequire.resolve(module);
      return module;
    } catch (e) {}
  }
  throw new Error(`Docusaurus was unable to resolve the "${moduleName}" ${moduleType}. Make sure one of the following packages are installed:
  ${modulePatterns.map((module) => `- ${module}`).join('\n')}`);
}
