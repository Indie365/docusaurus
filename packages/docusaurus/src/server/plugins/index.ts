/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {generate} from '@docusaurus/utils';
import fs from 'fs-extra';
import path from 'path';
import {
  LoadContext,
  PluginConfig,
  PluginContentLoadedActions,
  RouteConfig,
} from '@docusaurus/types';
import initPlugins, {PluginWithVersionInformation} from './init';
import chalk from 'chalk';

export function sortConfig(routeConfigs: RouteConfig[]): void {
  // Sort the route config. This ensures that route with nested
  // routes is always placed last.
  routeConfigs.sort((a, b) => {
    if (a.routes && !b.routes) {
      return 1;
    }
    if (!a.routes && b.routes) {
      return -1;
    }
    // Higher priority get placed first.
    if (a.priority || b.priority) {
      const priorityA = a.priority || 0;
      const priorityB = b.priority || 0;
      const score = priorityB - priorityA;

      if (score !== 0) {
        return score;
      }
    }

    return a.path.localeCompare(b.path);
  });

  routeConfigs.forEach((routeConfig) => {
    routeConfig.routes?.sort((a, b) => a.path.localeCompare(b.path));
  });
}

export function warnAboutOverridingRoutes(
  pluginsRouteConfigs: RouteConfig[],
): void {
  // Accumulate all the routes by recursively exploring each RouteConfig
  const routesAccumulator: string[] = [];
  function getAllRoutes(routeConfigs: RouteConfig[]): string[] {
    for (let i = 0; i < routeConfigs.length; i += 1) {
      const routeConfig = routeConfigs[i];
      routesAccumulator.push(routeConfig.path);
      if (routeConfig.routes !== undefined) {
        getAllRoutes(routeConfig.routes);
      }
    }
    return routesAccumulator;
  }

  const allRoutes = getAllRoutes(pluginsRouteConfigs);

  // Sort the allRoutes array in lexicographical order
  // Then check if each route is equal to the next route
  // If yes, one of these routes will be overridden so we warn the user
  allRoutes.sort((a, b) => a.localeCompare(b));
  for (let i = 0; i < allRoutes.length - 1; i += 1) {
    if (allRoutes[i] === allRoutes[i + 1]) {
      console.warn(
        `${
          chalk.yellow(`warning `) + chalk.bold.yellow(`Routes Override: `)
        }Attempting to create page at "${
          allRoutes[i]
        }" but a page already exists at this path\n${chalk.bold.yellow(
          `This could lead to non-deterministic routing behavior`,
        )}`,
      );
    }
  }
}

export async function loadPlugins({
  pluginConfigs,
  context,
}: {
  pluginConfigs: PluginConfig[];
  context: LoadContext;
}): Promise<{
  plugins: PluginWithVersionInformation[];
  pluginsRouteConfigs: RouteConfig[];
}> {
  // 1. Plugin Lifecycle - Initialization/Constructor.
  const plugins: PluginWithVersionInformation[] = initPlugins({
    pluginConfigs,
    context,
  });

  // 2. Plugin Lifecycle - loadContent.
  // Currently plugins run lifecycle methods in parallel and are not order-dependent.
  // We could change this in future if there are plugins which need to
  // run in certain order or depend on others for data.
  const pluginsLoadedContent = await Promise.all(
    plugins.map(async (plugin) => {
      if (!plugin.loadContent) {
        return null;
      }

      return plugin.loadContent();
    }),
  );

  // 3. Plugin Lifecycle - contentLoaded.
  const pluginsRouteConfigs: RouteConfig[] = [];

  await Promise.all(
    plugins.map(async (plugin, index) => {
      if (!plugin.contentLoaded) {
        return;
      }

      const pluginContentDir = path.join(
        context.generatedFilesDir,
        plugin.name,
      );

      const actions: PluginContentLoadedActions = {
        addRoute: (config) => pluginsRouteConfigs.push(config),
        createData: async (name, content) => {
          const modulePath = path.join(pluginContentDir, name);
          await fs.ensureDir(path.dirname(modulePath));
          await generate(pluginContentDir, name, content);
          return modulePath;
        },
      };

      await plugin.contentLoaded({
        content: pluginsLoadedContent[index],
        actions,
      });
    }),
  );

  warnAboutOverridingRoutes(pluginsRouteConfigs);

  // 4. Plugin Lifecycle - routesLoaded.
  // Currently plugins run lifecycle methods in parallel and are not order-dependent.
  // We could change this in future if there are plugins which need to
  // run in certain order or depend on others for data.
  await Promise.all(
    plugins.map(async (plugin) => {
      if (!plugin.routesLoaded) {
        return null;
      }

      return plugin.routesLoaded(pluginsRouteConfigs);
    }),
  );

  // Sort the route config. This ensures that route with nested
  // routes are always placed last.
  sortConfig(pluginsRouteConfigs);

  return {
    plugins,
    pluginsRouteConfigs,
  };
}
