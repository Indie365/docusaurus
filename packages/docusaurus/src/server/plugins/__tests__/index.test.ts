/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';

import {loadContext} from '@docusaurus/core/src/server/index';

describe('functional plugin', () => {
  async function loadSite() {
    const siteDir = path.join(__dirname, '__fixtures__', 'site-with-plugin');
    const context = await loadContext(siteDir);

    return {siteDir, context};
  }

  test('plugin gets parsed correctly and loads', async () => {
    const {context} = await loadSite();
    expect(context.siteConfig.plugins.length).toBe(3);
  });

  test('plugin should load in order', async () => {
    const {context} = await loadSite();
    const {plugins} = context.siteConfig;
    expect(
      typeof plugins[0] === 'function' ? plugins[0](context, {}).name : null,
    ).toBe('first-plugin');
    expect(
      typeof plugins[1][0] === 'function'
        ? plugins[1][0](context, {}).name
        : null,
    ).toBe('second-plugin');
  });
});
