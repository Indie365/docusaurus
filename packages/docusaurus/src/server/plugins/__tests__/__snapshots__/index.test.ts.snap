// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadPlugins loads plugins 1`] = `
{
  "globalData": {
    "test1": {
      "default": {
        "content": "a",
        "prop": "a",
      },
    },
  },
  "plugins": [
    {
      "content": "a",
      "contentLoaded": [Function],
      "loadContent": [Function],
      "name": "test1",
      "options": {
        "id": "default",
      },
      "path": "<PROJECT_ROOT>/packages/docusaurus/src/server/plugins/__tests__/__fixtures__/site-with-plugin",
      "prop": "a",
      "version": {
        "type": "local",
      },
    },
    {
      "configureWebpack": [Function],
      "content": undefined,
      "name": "test2",
      "options": {
        "id": "default",
      },
      "path": "<PROJECT_ROOT>/packages/docusaurus/src/server/plugins/__tests__/__fixtures__/site-with-plugin",
      "version": {
        "type": "local",
      },
    },
    {
      "content": undefined,
      "getClientModules": [Function],
      "injectHtmlTags": [Function],
      "name": "docusaurus-bootstrap-plugin",
      "options": {
        "id": "default",
      },
      "path": "<PROJECT_ROOT>/packages/docusaurus/src/server/plugins/__tests__/__fixtures__/site-with-plugin",
      "version": {
        "type": "synthetic",
      },
    },
    {
      "configureWebpack": [Function],
      "content": undefined,
      "name": "docusaurus-mdx-fallback-plugin",
      "options": {
        "id": "default",
      },
      "path": ".",
      "version": {
        "type": "synthetic",
      },
    },
  ],
  "pluginsRouteConfigs": [
    {
      "component": "Comp",
      "context": {
        "data": {
          "content": "path",
        },
        "plugin": "<PROJECT_ROOT>/packages/docusaurus/src/server/plugins/__tests__/__fixtures__/site-with-plugin/.docusaurus/test1/default/plugin-route-context-module-100.json",
      },
      "lastModified": 2023-08-03T00:00:00.000Z,
      "modules": {
        "content": "path",
      },
      "path": "foo/",
    },
  ],
}
`;
